### PR TITLE
Implement jet ID for PUPPI jets

### PR DESCRIPTION
**Pull request by ahinzmann**
This implements jet ID for PUPPI jets and applies it in Offline DQM.

In CMSSW>=10_1_X this should have a small effect on the CleanedslimmedJetsPuppi DQM plots, since the jet ID is based on puppiMultiplicties available in MiniAOD.

(In CMSSW_9_4_X this would remove a large fraction of forward jets in the CleanedslimmedJetsPuppi DQM plots if #23012 is not applied. If #23012 is applied on top, this should have a small effect.)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@cmsbuild please test
(relvals timed out: let try again)

File: PhysicsTools/SelectorUtils/interface/PFJetIDSelectionFunctor.h [hunk 1]
```diff
@@ -243,40 +282,51 @@ class PFJetIDSelectionFunctor : public Selector<pat::Jet>  {
 
     push_back("CHF" );
     push_back("NHF" );
```
Comment (slava77) [inline PhysicsTools/SelectorUtils/interface/PFJetIDSelectionFunctor.h]: for a future PR I made #23131 to hopefully reduce the duplicated implementation of the cuts

**Comment by ahinzmann** [comment]
hold.
According to this plot, the change does not seem to be picked up in this DQM plot:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-05-02-1100+23120/26370/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/JetMET__Jet_CleanedslimmedJetsPuppi_Constituents_Forward.png
The new cut on number of constituents for PUPPI jets is 2<nconstituents<10 for eta>3. I'm not sure which config one would need to modify for this DQM plot.

**Comment by ahinzmann** [comment]
I removed the code duplication to address #23131. However, I didn't understand yet why the jet ID change does not show up in the above mentioned DQM plot. Any advice?

File: DQMOffline/JetMET/python/jetAnalyzer_cfi.py [hunk 2]
```diff
@@ -164,12 +164,20 @@
 
 jetDQMAnalyzerAk8PFPUPPICleanedMiniAOD=jetDQMAnalyzerAk4PFCHSCleanedMiniAOD.clone(
     jetsrc = cms.InputTag("slimmedJetsAK8"),
```
Comment (ahinzmann) [inline DQMOffline/JetMET/python/jetAnalyzer_cfi.py]: @ramankhurana  shouldn't there also be a line "JetType = cms.string('miniaod')," here?

Comment (ramankhurana) [reply DQMOffline/JetMET/python/jetAnalyzer_cfi.py]: @ahinzmann  Yes it should be there, I see the link from your PR:

https://github.com/ahinzmann/cmssw/blob/fa8a49d9e88f4b9d3b570f935eda196593105d7a/DQMOffline/JetMET/python/jetAnalyzer_cfi.py

that this line is already there. I am not sure what is the confusion/problem. Please let me know.

Comment (ahinzmann) [reply DQMOffline/JetMET/python/jetAnalyzer_cfi.py]: ok, yes, it is there

**Comment by dmitrijus** [comment]
+1

**Comment by slava77** [comment]
@ahinzmann 
please clarify if you expect more changes in this PR.

**Comment by ahinzmann** [comment]
I would like to understand why the change is not visible in this baseline comparison before merging:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-05-02-1100+23120/26370/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/JetMET__Jet_CleanedslimmedJetsPuppi_Constituents_Forward.png
I have no clue at the moment

**Comment by ahinzmann** [comment]
Ok, the above described feature is understood. I've adapted the DQM plots to use the same definition of nConstituents for PUPPI jets as the jet ID.

**Comment by perrotta** [comment]
please test

**Comment by ahinzmann** [comment]
This behaves as expected now:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_2_X_2018-05-07-2300+23120/26431/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/JetMET__Jet_CleanedslimmedJetsPuppi_Constituents_Forward.png
Ready to go from my point of view.

**Comment by dmitrijus** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23120	b6c9039
- technical changes since the last reco signoff in reco-related code PhysicsTools/SelectorUtils (#23120	0c0f2e7); main changes in the last update were in the DQM code
- jenkins tests pass and comparisons with the baseline show differences only in the DQM monitoring, including folders Jet/CleanedslimmedJetsPuppi\|Jet/CleanedslimmedJetsAK8\|MET/slimmedMETsPuppi. The changes are apparently expected and primarily affect the forward regions, e.g. wf 136.8311 
wf136 8311_puppi_constforward

**Comment by slava77** [comment]
@fabiocos 
please clarify if there are some issues with this PR.
Thank you.

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**